source/common/buffer/buffer_impl.h

```diff
@@ -1,16 +1,98 @@
 #pragma once
 
 #include <cstdint>
+#include <deque>
 #include <string>
 
 #include "envoy/buffer/buffer.h"
 
 #include "common/common/non_copyable.h"
-#include "common/event/libevent.h"
 
 namespace Envoy {
 namespace Buffer {
 
+/**
+ * A BufferSlice manages a contiguous block of bytes.
+ * The block is arranged like this:
+ *                   |<- data_size() -->|                    |<- reservable_size() ->|
+ * +-----------------+------------------+--------------------+-----------------------+
+ * | Drained         | Data             | Reserved           | Reservable            |
+ * | Unused space    | Usable content   | Additional content | Unused space that can |
+ * | that formerly   |                  | can be added here  | can be moved into     |
+ * | was in the Data |                  |                    | the Reserved section  |
+ * | section         |                  |                    |                       |
+ * +-----------------+------------------+--------------------+-----------------------+
+ *                   ^
+ *                   |
+ *                   data()
+ */
+class BufferSlice {
+public:
+  using Reservation = std::pair<void*, uint64_t>;
+
+  virtual ~BufferSlice() = default;
+
+  /**
+   * @return a pointer to the start of the usable content.
+   */
+  virtual const void* data() const PURE;
+
+  /**
+   * @return a pointer to the start of the usable content.
+   */
+  virtual void* data() PURE;
+
+  /**
+   * @return the size in bytes of the usable content.
+   */
+  virtual uint64_t dataSize() const PURE;
+
+  /**
+   * Remove the first `size` bytes of usable content. Runs in O(1) time.
+   * @param size number of bytes to remove. If greater than data_size(), the result is undefined.
+   */
+  virtual void drain(uint64_t size) PURE;
+
+  /**
+   * @return the number of bytes available to be reserve()d.
+   * @note Read-only implementations of BufferSlice should return zero from this method.
+   */
+  virtual uint64_t reservableSize() const PURE;
+
+  /**
+   * Reserve `size` bytes that the caller can populate with content. The caller SHOULD then
+   * call commit() to add the newly populated content from the Reserved section to the Data
+   * section.
+   * @note It is valid to call reserve() multiple times without a commit in between. The
+   *       result will be a succession of contiguous, reserved spans of memory, where the
+   *       first span starts right after the Data section. These spans MUST be commit()ed
+   *       in the order in which they were reserved; the result of calling commit() out
+   *       of order is undefined.
+   * @param size the number of bytes to reserve. The BufferSlice implementation MAY reserve
+   *        fewer bytes than requested (for example, if it doesn't have enough room in the
+   *        Reservable section to fulfill the whole request).
```

Comment:
so it's the responsibility of the caller to check the reservation-size and create a new slice if needed?

Reply:
Almost -- It's the responsibility of the caller to check the reservation size and obtain the remaining from another slice if needed. That might mean creating a new slice, but it could also mean reusing some other existing slice (which `OwnedImpl` does when possible).